From 29fef8ff70b2551961990bc23325f3269fc75246 Mon Sep 17 00:00:00 2001
From: Qiu Yuzhou <qiuyuzhou@gmail.com>
Date: Tue, 19 Apr 2022 11:28:13 +0800
Subject: [PATCH] chore(chart): update ingress to support apiVersion
 networking.k8s.io/v1 (#1769)

* chore(chart): update ingress to support apiVersion networking.k8s.io/v1

* refactor(chart): use API capabilities instead version comparisions in ingress

Co-authored-by: Manu Gupta <manugupt1@gmail.com>
---
 charts/athens-proxy/Chart.yaml             |  4 +-
 charts/athens-proxy/templates/ingress.yaml | 79 ++++++++++++++++------
 charts/athens-proxy/values.yaml            |  5 ++
 3 files changed, 64 insertions(+), 24 deletions(-)

diff --git a/charts/athens-proxy/Chart.yaml b/charts/athens-proxy/Chart.yaml
index 7df2e024..c2b54b67 100644
--- a/charts/athens-proxy/Chart.yaml
+++ b/charts/athens-proxy/Chart.yaml
@@ -1,7 +1,7 @@
 apiVersion: v1
 name: athens-proxy
-version: 0.5.0
-appVersion: 0.11.0
+version: 0.5.2
+appVersion: 0.11.1
 description: The proxy server for Go modules
 icon: https://raw.githubusercontent.com/gomods/athens/main/docs/static/banner.png
 keywords:
diff --git a/charts/athens-proxy/templates/ingress.yaml b/charts/athens-proxy/templates/ingress.yaml
index d3303757..1fbd85cd 100644
--- a/charts/athens-proxy/templates/ingress.yaml
+++ b/charts/athens-proxy/templates/ingress.yaml
@@ -1,43 +1,78 @@
 {{- if .Values.ingress.enabled -}}
-{{- $serviceName := include "fullname" . -}}
-{{- $servicePort := .Values.service.servicePort -}}
-{{- if .Capabilities.APIVersions.Has "networking.k8s.io/v1beta1" -}}
+{{- $fullName := include "fullname" . -}}
+{{- $svcPort := .Values.service.servicePort -}}
+{{- if and .Values.ingress.className (not (semverCompare ">=1.18-0" .Capabilities.KubeVersion.GitVersion)) }}
+  {{- if not (hasKey .Values.ingress.annotations "kubernetes.io/ingress.class") }}
+  {{- $_ := set .Values.ingress.annotations "kubernetes.io/ingress.class" .Values.ingress.className}}
+  {{- end }}
+{{- end }}
+{{- if .Capabilities.APIVersions.Has "networking.k8s.io/v1" -}}
+apiVersion: networking.k8s.io/v1
+{{- else if .Capabilities.APIVersions.Has "networking.k8s.io/v1beta1" -}}
 apiVersion: networking.k8s.io/v1beta1
 {{- else -}}
 apiVersion: extensions/v1beta1
 {{- end }}
 kind: Ingress
 metadata:
-  name: {{ template "fullname" . }}
+  name: {{ $fullName }}
   labels:
     app: {{ template "fullname" . }}
     chart: {{ .Chart.Name }}-{{ .Chart.Version | replace "+" "_" }}
     release: {{ .Release.Name }}
     heritage: {{ .Release.Service }}
+  {{- with .Values.ingress.annotations }}
   annotations:
-    {{- range $key, $value := .Values.ingress.annotations }}
-      {{ $key }}: {{ $value | quote }}
-    {{- end }}
+    {{- toYaml . | nindent 4 }}
+  {{- end }}
 spec:
-  backend:
-    serviceName: {{ $serviceName }}
-    servicePort: {{ $servicePort }}
+  {{- if and .Values.ingress.className (semverCompare ">=1.18-0" .Capabilities.KubeVersion.GitVersion) }}
+  ingressClassName: {{ .Values.ingress.className }}
+  {{- end }}
+  {{- if .Values.ingress.tls }}
+  tls:
+    {{- range .Values.ingress.tls }}
+    - hosts:
+        {{- range .hosts }}
+        - {{ . | quote }}
+        {{- end }}
+      secretName: {{ .secretName }}
+    {{- end }}
+  {{- end }}
   rules:
-    {{- range $host := .Values.ingress.hosts }}
-    - host: {{ $host }}
+    {{- range .Values.ingress.hosts }}
+    - host: {{ .host | quote }}
       http:
         paths:
+          {{- range .paths }}
+          - path: {{ .path }}
+            {{- if and .pathType (semverCompare ">=1.18-0" $.Capabilities.KubeVersion.GitVersion) }}
+            pathType: {{ .pathType }}
+            {{- end }}
+            backend:
+              {{- if semverCompare ">=1.19-0" $.Capabilities.KubeVersion.GitVersion }}
+              service:
+                name: {{ $fullName }}
+                port:
+                  number: {{ $svcPort }}
+              {{- else }}
+              serviceName: {{ $fullName }}
+              servicePort: {{ $svcPort }}
+              {{- end }}
+          {{- end }}
           - path: /
+            {{- if and .pathType (semverCompare ">=1.18-0" $.Capabilities.KubeVersion.GitVersion) }}
+            pathType: {{ .pathType }}
+            {{- end }}
             backend:
+              {{- if semverCompare ">=1.19-0" $.Capabilities.KubeVersion.GitVersion }}
+              service:
+                name: ssl-redirect
+                port:
+                  number: use-annotation
+              {{- else }}
               serviceName: ssl-redirect
               servicePort: use-annotation
-          - path: /
-            backend:
-              serviceName: {{ $serviceName }}
-              servicePort: {{ $servicePort }}
-    {{- end -}}
-  {{- if .Values.ingress.tls }}
-  tls:
-{{ toYaml .Values.ingress.tls | indent 4 }}
-  {{- end -}}
-{{- end -}}
\ No newline at end of file
+              {{- end }}
+    {{- end }}
+{{- end }}
diff --git a/charts/athens-proxy/values.yaml b/charts/athens-proxy/values.yaml
index e4c20962..b60f213d 100644
--- a/charts/athens-proxy/values.yaml
+++ b/charts/athens-proxy/values.yaml
@@ -39,8 +39,13 @@ ingress:
   enabled: false
   # Provide key/value annotations
   annotations:
+  className: ""
   # Provide an array of values for the ingress host mapping
   hosts:
+    # - host: athens-proxy.local
+    # paths:
+    #   - path: /
+    #     pathType: ImplementationSpecific
   # Provide a base64 encoded cert for TLS use
   tls:
 
-- 
2.39.2

